### ISSUE TITLE
Input image dtype is bool. Interpolation is not defined with bool data type.

### ISSUE THREAD

**Issue by parth-singh71** [matterport/Mask_RCNN#2243]
I am facing this issue while using Mask_RCNN to train on my custom dataset with multiple classes.

This error occurs when I start training. This is what I get:

`/home/parth/anaconda3/envs/compVision/lib/python3.7/site-packages/skimage/transform/_warps.py:830: FutureWarning: Input image dtype is bool. Interpolation is not defined with bool data type. Please set order to 0 or explicitely cast input image to another data type. Starting from version 0.19 a ValueError will be raised instead of this warning.
  order = _validate_interpolation_order(image.dtype, order)`

I keep getting this for like a hundred times and then the kernel dies.
Please Help!!

**Comment by naaths** [comment]
Hello, have you solved it

**Comment by naaths** [comment]
try the skimage version 0.16.2

**Comment by parth-singh71** [comment]
Yeah I solved it. Apart from the changes you mentioned above I also had to downgrade to:
```
tensorflow=1.5.0
keras=2.1.5
```
After these changes it was working fine.

**Comment by devloper13** [comment]
> I am facing this issue while using Mask_RCNN to train on my custom dataset with multiple classes.
> 
> This error occurs when I start training. This is what I get:
> 
> `/home/parth/anaconda3/envs/compVision/lib/python3.7/site-packages/skimage/transform/_warps.py:830: FutureWarning: Input image dtype is bool. Interpolation is not defined with bool data type. Please set order to 0 or explicitely cast input image to another data type. Starting from version 0.19 a ValueError will be raised instead of this warning. order = _validate_interpolation_order(image.dtype, order)`
> 
> I keep getting this for like a hundred times and then the kernel dies.
> Please Help!!

This was a warning, wasn't it? Did it effect your training?

**Comment by SanjaiPraboo** [comment]
I am also getting the same warning, but training wasn't affected

**Comment by panovr** [comment]
I just wonder if we want to modify code to eliminate this warning, where and how should we modify?

**Comment by Atonbom** [comment]
Did anyone fix this? I'm having the same "issue". My model is training but I'm getting a load of these warnings

I use:
tensorflu-gpu 1.15
keras 2.08

I had LZW compression errors for my tiff files so I uninstalled the following packages and installed with these versions
scikit-image 0.18.1 
tifffile 2021.3.31
imagecodecs 2021.3.31

**Comment by ic** [comment]
## Update

The original post hides a problem. Here is a working version.

```
diff --git a/mrcnn/utils.py b/mrcnn/utils.py
index dd29453..3b7b760 100644
--- a/mrcnn/utils.py
+++ b/mrcnn/utils.py
@@ -18,6 +18,7 @@ import scipy
 import skimage.color
 import skimage.io
 import skimage.transform
+from skimage.util import img_as_float
 import urllib.request
 import shutil
 import warnings
@@ -895,19 +896,21 @@ def resize(image, output_shape, order=1, mode='constant', cval=0, clip=True,
     of skimage. This solves the problem by using different parameters per
     version. And it provides a central place to control resizing defaults.
     """
+    imgf = image.astype(np.float32)
     if LooseVersion(skimage.__version__) >= LooseVersion("0.14"):
         # New in 0.14: anti_aliasing. Default it to False for backward
         # compatibility with skimage 0.13.
         return skimage.transform.resize(
-            image, output_shape,
+            imgf, output_shape,
             order=order, mode=mode, cval=cval, clip=clip,
             preserve_range=preserve_range, anti_aliasing=anti_aliasing,
     else:
         return skimage.transform.resize(
-            image, output_shape,
+            imgf, output_shape,
             order=order, mode=mode, cval=cval, clip=clip,
```

The problem with the original change is that the Scikit utility `img_as_float` does not only convert the data type, [but also scales](https://scikit-image.org/docs/0.19.x/api/skimage.util.html#img-as-float) to [0, 1]. So training works, but it is all wrong, as the data values are normalised whereas the algorithm does not expect it.

---
## Original post
The problem is in `scikit-image` after `0.16.x` apparently. Here using `0.19.3` with the attached patch.

```
diff --git a/mrcnn/utils.py b/mrcnn/utils.py
index dd29453..3b7b760 100644
--- a/mrcnn/utils.py
+++ b/mrcnn/utils.py
@@ -18,6 +18,7 @@ import scipy
 import skimage.color
 import skimage.io
 import skimage.transform
+from skimage.util import img_as_float
 import urllib.request
 import shutil
 import warnings
@@ -895,19 +896,21 @@ def resize(image, output_shape, order=1, mode='constant', cval=0, clip=True,
     of skimage. This solves the problem by using different parameters per
     version. And it provides a central place to control resizing defaults.
     """
+    return_type = image.dtype
+    imgf = img_as_float(image)
     if LooseVersion(skimage.__version__) >= LooseVersion("0.14"):
         # New in 0.14: anti_aliasing. Default it to False for backward
         # compatibility with skimage 0.13.
         return skimage.transform.resize(
-            image, output_shape,
+            imgf, output_shape,
             order=order, mode=mode, cval=cval, clip=clip,
             preserve_range=preserve_range, anti_aliasing=anti_aliasing,
-            anti_aliasing_sigma=anti_aliasing_sigma)
+            anti_aliasing_sigma=anti_aliasing_sigma).astype(return_type)
     else:
         return skimage.transform.resize(
-            image, output_shape,
+            imgf, output_shape,
             order=order, mode=mode, cval=cval, clip=clip,
-            preserve_range=preserve_range)
+            preserve_range=preserve_range).astype(return_type)
```

(it could be a push-button patch to merge with GitHub, but working elsewhere~)

**Comment by MajidNikoogoftar** [comment]
@ Hello
It dosen't solve the problem

**Comment by ekremcet** [comment]
> The problem is in `scikit-image` after `0.16.x` apparently. Here using `0.19.3` with the attached patch.
> 
> ```
> diff --git a/mrcnn/utils.py b/mrcnn/utils.py
> index dd29453..3b7b760 100644
> --- a/mrcnn/utils.py
> +++ b/mrcnn/utils.py
> @@ -18,6 +18,7 @@ import scipy
>  import skimage.color
>  import skimage.io
>  import skimage.transform
> +from skimage.util import img_as_float
>  import urllib.request
>  import shutil
>  import warnings
> @@ -895,19 +896,21 @@ def resize(image, output_shape, order=1, mode='constant', cval=0, clip=True,
>      of skimage. This solves the problem by using different parameters per
>      version. And it provides a central place to control resizing defaults.
>      """
> +    return_type = image.dtype
> +    imgf = img_as_float(image)
>      if LooseVersion(skimage.__version__) >= LooseVersion("0.14"):
>          # New in 0.14: anti_aliasing. Default it to False for backward
>          # compatibility with skimage 0.13.
>          return skimage.transform.resize(
> -            image, output_shape,
> +            imgf, output_shape,
>              order=order, mode=mode, cval=cval, clip=clip,
>              preserve_range=preserve_range, anti_aliasing=anti_aliasing,
> -            anti_aliasing_sigma=anti_aliasing_sigma)
> +            anti_aliasing_sigma=anti_aliasing_sigma).astype(return_type)
>      else:
>          return skimage.transform.resize(
> -            image, output_shape,
> +            imgf, output_shape,
>              order=order, mode=mode, cval=cval, clip=clip,
> -            preserve_range=preserve_range)
> +            preserve_range=preserve_range).astype(return_type)
> ```
> 
> (it could be a push-button patch to merge with GitHub, but working elsewhere~)

Thank you, this solved the issue for me. (tested with `scikit-image` `0.19.1`)

**Comment by bruce-goldfeder-asetpartners** [comment]
With the patch the code functions, but it is not training correctly?  Tensorflow 1.15.5 using nvidia-index on 3090, cuda 11.5, nvidia driver 510

**Comment by ic** [comment]
@bruce-goldfeder-asetpartners Have you used the updated version (to be sure) ? My hardware settings are way older (CUDA 10.2), but same Tensorflow version.

It trains correctly here with the updated version of the patch. There was a training issue with the original one.

**Comment by nataliameira** [comment]
Hello @ic @bruce-goldfeder-asetpartners 

Can you show me your resize function?

**Comment by gbinduo** [comment]
> try the skimage version 0.16.2

Perfect! Thanks!

**Comment by LiuxinYLX** [comment]
Sorry to disturb you. Could I ask you if we should modify the utils.py in the address like this /python3.7/site-packages/mask_rcnn-2.1-py3.7.egg/mrcnn/ or modify the utils.py in Mask_RCNN/mrcnn ( where we execute the code), please? 

I modified the utils.py in Mask_RCNN/mrcnn. And I still have problems when I do like this : 



def resize(image, output_shape, order=1, mode='constant', cval=0, clip=True,
           preserve_range=False, anti_aliasing=False, anti_aliasing_sigma=None):
    """A wrapper for Scikit-Image resize().

    Scikit-Image generates warnings on every call to resize() if it doesn't
    receive the right parameters. The right parameters depend on the version
    of skimage. This solves the problem by using different parameters per
    version. And it provides a central place to control resizing defaults.
    """
    return_type = image.dtype
    imgf = img.astype(np.float32)

    if LooseVersion(skimage.__version__) >= LooseVersion("0.14"):
        # New in 0.14: anti_aliasing. Default it to False for backward
        # compatibility with skimage 0.13.
        return skimage.transform.resize(
            imgf, output_shape,
            order=order, mode=mode, cval=cval, clip=clip,
            preserve_range=preserve_range, anti_aliasing=anti_aliasing,
            anti_aliasing_sigma=anti_aliasing_sigma).astype(return_type)
    else:
        return skimage.transform.resize(
            imgf, output_shape,
            order=order, mode=mode, cval=cval, clip=clip,
            preserve_range=preserve_range).astype(return_type)

            
 Could you please help me?

**Comment by DiogoBotton** [comment]
> Sorry to disturb you. Could I ask you if we should modify the utils.py in the address like this /python3.7/site-packages/mask_rcnn-2.1-py3.7.egg/mrcnn/ or modify the utils.py in Mask_RCNN/mrcnn ( where we execute the code), please?
> 
> I modified the utils.py in Mask_RCNN/mrcnn. And I still have problems when I do like this :
> 
> def resize(image, output_shape, order=1, mode='constant', cval=0, clip=True, preserve_range=False, anti_aliasing=False, anti_aliasing_sigma=None): """A wrapper for Scikit-Image resize().
> 
> ```
> Scikit-Image generates warnings on every call to resize() if it doesn't
> receive the right parameters. The right parameters depend on the version
> of skimage. This solves the problem by using different parameters per
> version. And it provides a central place to control resizing defaults.
> """
> return_type = image.dtype
> imgf = img.astype(np.float32)
> 
> if LooseVersion(skimage.__version__) >= LooseVersion("0.14"):
>     # New in 0.14: anti_aliasing. Default it to False for backward
>     # compatibility with skimage 0.13.
>     return skimage.transform.resize(
>         imgf, output_shape,
>         order=order, mode=mode, cval=cval, clip=clip,
>         preserve_range=preserve_range, anti_aliasing=anti_aliasing,
>         anti_aliasing_sigma=anti_aliasing_sigma).astype(return_type)
> else:
>     return skimage.transform.resize(
>         imgf, output_shape,
>         order=order, mode=mode, cval=cval, clip=clip,
>         preserve_range=preserve_range).astype(return_type)
> ```
> 
> Could you please help me?

I was trying to modify the code in the utils.py file, but I was facing the same issue as you. When I ran the training, it seemed like none of the changes I made to the code were taking effect. After a long night, I discovered that the problem was the cache of Visual Studio Code. Once I deleted the cache, everything worked like magic (I found this out after 4 hours). As always, cache has always been and still is the biggest enemy of programmers haha.

However, as a colleague mentioned, changing the skimage version will only be a temporary band-aid.

The error itself suggests that we need to change the data type to boolean. One way to do this is by setting the order parameter to 0 (zero) when the image is a boolean array and to 1 when it’s a float array. Basically, like this:

```py
if image.dtype == bool:
    order = 0
```

The final code will look like this:

```py
def resize(image, output_shape, order=1, mode='constant', cval=0, clip=True,
           preserve_range=False, anti_aliasing=False, anti_aliasing_sigma=None):
    """A wrapper for Scikit-Image resize().

    Scikit-Image generates warnings on every call to resize() if it doesn't
    receive the right parameters. The correct parameters depend on the skimage version. 
    This solves the problem by using different parameters for each version. 
    It also provides a central place to control resizing defaults.
    """
    if image.dtype == bool:
        order = 0

    if LooseVersion(skimage.__version__) >= LooseVersion("0.14"):
        # New in 0.14: anti_aliasing. Default it to False for backward
        # compatibility with skimage 0.13.
        return skimage.transform.resize(
            image, output_shape,
            order=order, mode=mode, cval=cval, clip=clip,
            preserve_range=preserve_range, anti_aliasing=anti_aliasing,
            anti_aliasing_sigma=anti_aliasing_sigma)
    else:
        return skimage.transform.resize(
            image, output_shape,
            order=order, mode=mode, cval=cval, clip=clip,
            preserve_range=preserve_range)
```

This worked for me. I hope this helps at least some of you avoid staying up all night fixing bugs hahaha!
